File: _python_backup/services/file_service.py
"""
File handling service
"""

from typing import List, Optional
from domain.models import FileUploadInfo
from domain.exceptions import FileError, ValidationError
from config.constants import ALLOWED_FILE_TYPES, MAX_FILE_SIZE_MB


class FileService:
    """Service for file operations"""
    
    @staticmethod
    def validate_file(file_name: str, file_size: int) -> None:
        """
        Validate uploaded file
        
        Args:
            file_name: Name of the file
            file_size: Size of the file in bytes
            
        Raises:
            ValidationError: If file is invalid
        """
        # Check file extension
        if not any(file_name.lower().endswith(ext) for ext in ALLOWED_FILE_TYPES):
            raise ValidationError(
                f"File type not allowed. Allowed types: {', '.join(ALLOWED_FILE_TYPES)}"
            )
        
        # Check file size
        size_mb = file_size / (1024 * 1024)
        if size_mb > MAX_FILE_SIZE_MB:
            raise ValidationError(
                f"File size ({size_mb:.2f} MB) exceeds maximum allowed size "
                f"({MAX_FILE_SIZE_MB} MB)"
            )
    
    @staticmethod
    def read_file_content(uploaded_file) -> str:
        """
        Read content from uploaded file
        
        Args:
            uploaded_file: Streamlit uploaded file object
            
        Returns:
            File content as string
            
        Raises:
            FileError: If reading fails
        """
        try:
            return uploaded_file.read().decode('utf-8')
        except Exception as e:
            raise FileError(f"Failed to read file {uploaded_file.name}: {str(e)}") from e
    
    @staticmethod
    def get_file_info(uploaded_file) -> FileUploadInfo:
        """
        Get file information
        
        Args:
            uploaded_file: Streamlit uploaded file object
            
        Returns:
            FileUploadInfo domain model
        """
        return FileUploadInfo(
            name=uploaded_file.name,
            size_bytes=len(uploaded_file.getvalue())
        )

File: js/src/services/fileService.js
/**
 * File handling service - JavaScript version
 */

const { FileUploadInfo } = typeof module !== 'undefined' && module.exports
    ? require('../domain/models')
    : window.VNGDomain;

const { FileError, ValidationError } = typeof module !== 'undefined' && module.exports
    ? require('../domain/exceptions')
    : window.VNGExceptions;

const { ALLOWED_FILE_TYPES, MAX_FILE_SIZE_MB } = typeof module !== 'undefined' && module.exports
    ? require('../config/constants')
    : window.VNGConstants;

/**
 * Service for file operations
 */
class FileService {
    /**
     * Validate uploaded file
     * @param {string} fileName - Name of the file
     * @param {number} fileSize - Size of the file in bytes
     */
    static validateFile(fileName, fileSize) {
        // Check file extension
        const isAllowedType = ALLOWED_FILE_TYPES.some(ext =>
            fileName.toLowerCase().endsWith(ext)
        );

        if (!isAllowedType) {
            throw new ValidationError(
                `File type not allowed. Allowed types: ${ALLOWED_FILE_TYPES.join(', ')}`
            );
        }

        // Check file size
        const sizeMB = fileSize / (1024 * 1024);
        if (sizeMB > MAX_FILE_SIZE_MB) {
            throw new ValidationError(
                `File size (${sizeMB.toFixed(2)} MB) exceeds maximum allowed size (${MAX_FILE_SIZE_MB} MB)`
            );
        }
    }

    /**
     * Read content from uploaded file
     * @param {File} uploadedFile - File object from input
     * @returns {Promise<string>} File content as string
     */
    static async readFileContent(uploadedFile) {
        return new Promise((resolve, reject) => {
            const reader = new FileReader();

            reader.onload = (event) => {
                try {
                    resolve(event.target.result);
                } catch (error) {
                    reject(new FileError(`Failed to read file ${uploadedFile.name}: ${error.message}`));
                }
            };

            reader.onerror = () => {
                reject(new FileError(`Failed to read file ${uploadedFile.name}`));
            };

            reader.readAsText(uploadedFile);
        });
    }

    /**
     * Get file information
     * @param {File} uploadedFile - File object from input
     * @returns {FileUploadInfo} FileUploadInfo domain model
     */
    static getFileInfo(uploadedFile) {
        return new FileUploadInfo(
            uploadedFile.name,
            uploadedFile.size
        );
    }
}

// Export service
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { FileService };
} else if (typeof window !== 'undefined') {
    window.VNGServices = window.VNGServices || {};
    window.VNGServices.FileService = FileService;
}
